### ISSUE TITLE
How can I Improve on performance

### ISSUE THREAD

**Issue by zeedevio** [astronomer/astronomer-cosmos#840]
I'm using [Astro Airflow](https://docs.astronomer.io/astro/cli/overview) locally. 

So I'm trying to execute DBT transformation models using airflow and find it to be extremely slow.
I'm looking for help to improve the performance of my airflow dag. 

**I'm trying to run 56 transformation models, I have done the following test:**

- **DBT cloud:** it completes my model transformation in **4 minutes**
- **Testing Profile Locally**: using ` dbt run --profiles-dir /usr/local/airflow/include/dbt/` it completes my transformations in **6 minutes**
![Screenshot 2024-02-07 at 11 40 57](https://github.com/astronomer/astronomer-cosmos/assets/74731090/d2a4d42d-3551-4f53-88ce-4b8a15cd569f)

- **AIRFLOW:**: it takes  **44 minutes** and i get a fail
<img width="983" alt="Screenshot 2024-02-07 at 11 46 36" src="https://github.com/astronomer/astronomer-cosmos/assets/74731090/9e2c96d2-d2f4-429f-b25a-1eefeb454a29">

I'm looking for a way to run my models much faster.

**Comment by kzajaczkowski** [comment]
Having very similar performance issue.

**Comment by tatiana** [comment]
Some progress: #1014.

**Comment by zeedevio** [comment]
Hey Good day, 
Sorry for the late reply but I was able to speed up my transformation workflow models.

In my case I used GCP - Composer. Things I done to speed up transformations was to update Airflow Configuration:


Configuration | New Value
-- | --
dagbag_import_timeout | 5000
Parallelism | 32
max_active_task_per_dag | 16

**Comment by tatiana** [comment]
PR #1014 is significantly improving the performance with `LoadMode.DBT_LS`.

**Comment by tatiana** [comment]
If using `LoadMode.DBT_LS`, please, could you try Cosmos 1.5.0a9, which will be released as a stable version this week?
* https://github.com/astronomer/astronomer-cosmos/releases/tag/astronomer-cosmos-v1.5.0a9
* https://pypi.org/project/astronomer-cosmos/1.5.0a9/

Some ways to improve the performance using Cosmos 1.4:

**1. Can you pre-compile your dbt project?**

If yes, this would remove this responsibility from the Airflow DAG processor, greatly impacting the DAG parsing time. You could try this by using and specifying the path to the manifest file:
```
DbtDag(
    ...,
    render_config=RenderConfig(
        load_method=LoadMode.DBT_MANIFEST
    )
)
```

More information: https://astronomer.github.io/astronomer-cosmos/configuration/parsing-methods.html#dbt-manifest

**2. If you need to use `LoadMode.DBT_LS`, can you pre-install dbt dependencies in the Airflow scheduler and worker nodes?**

If yes, this will avoid Cosmos having to run dbt deps all the time before running any dbt command, both in the scheduler and worker nodes. In that case, you should set:
```
DbtDag(
    ...,
    operator_args={"install_deps": False}
    render_config=RenderConfig(
        dbt_deps=False
    )
)
```
More info:
https://astronomer.github.io/astronomer-cosmos/configuration/render-config.html

**3. If you need to use LoadMode.DBT_LS, is your dbt project large? Could you use selectors to select a subset?**

```
jaffle_shop = DbtDag(
    render_config=RenderConfig(
        select=["path:analytics"],
    )
)
```
More info: https://astronomer.github.io/astronomer-cosmos/configuration/selecting-excluding.html

**Comment by kzajaczkowski** [comment]
@tatiana, thank you for detailed explanations. They're helpful. Is using the manifest a recommended mode in terms of performance? If so will it still be the case after 1.5.0 is released?

**Comment by tatiana** [comment]
@kzajaczkowski, it really depends on your team's needs!

Using manifest is the safest approach in production from the perspective you're fully off-loading the Airflow DAG processor from ever having to run the `dbt ls` command. As you know, the `dbt ls` command can take some time to run, and it is proportional to the size of the project.

We understand that it is handy not to have to pre-compile the dbt project for many teams. If it is acceptable for the team that from time to time, the Airflow DAG processor will have to run `dbt ls` and that the cache will be stored in Airflow Variables, then performance-wise, the 1.5 release is the best performance we managed to achieve with `LoadMode.DBT_LS`.

In the past weeks, we've collaborated with customers who tested every iteration of the 1.5 alphas. They will deploy them live once we release the production version, giving us confidence in the stability of what we've built.
